Fix random forest lack of randomness (#1887)

rcurtin opened the pull request:
I had a chance to look into it today and I actually found two bugs in the random forest:

 * The decision tree was not properly initializing the dimension selector in the case of numeric-only data.
 * The random forest was not giving bootstrapped data to the decision trees.

I added a test to check that dimensions are selected randomly, and adapted the other tests (since bootstrapped data no longer gives the same guarantees as we have when we train directly on the full dataset).

gmanlan commented:
I tested the changes using the sample app, but the trained models are exactly the same every time you run the app. @zoq suggested to add mlpack::math::RandomSeed(time(NULL)); at the beginning of the program and by doing so it now generates different models every time.
So the question would be: shouldn't be the math::random initialization an intrinsic part of Random Forest?

zoq commented on src/mlpack/methods/decision_tree/decision_tree.hpp:
```diff
@@ -363,6 +363,10 @@ class DecisionTree :
   //! Modify the child of the given index (be careful!).
   DecisionTree& Child(const size_t i) { return *children[i]; }
 
+  //! Get the split dimension (only meaningful is this is a non-leaf in a
```
Should this be "if this is a ..."?

zoq commented on src/mlpack/tests/random_forest_test.cpp:
```diff
@@ -149,7 +149,7 @@ BOOST_AUTO_TEST_CASE(UnweightedNumericLearningTest)
   size_t dtCorrect = arma::accu(dtPredictions == testLabels);
 
-  BOOST_REQUIRE_GE(rfCorrect, dtCorrect);
+  BOOST_REQUIRE_GE(rfCorrect, dtCorrect * 0.85);
```
Did the performance in terms of accuracy drop with the changes?

gmanlan replied:
Using the sample app, I noticed the accuracy dropped around 10% in average... not sure about this test

zoq replied:
Thanks for the input, didn't expect to see 100% accuracy; I guess if we tweak the model parameter we can probably get the same result.

rcurtin replied:
Yeah, this was a bit fishy, and after a lot of investigation my conclusions:

 * The last change actually did introduce some problems.  I hope they are fixed now.
 * I'm not sure why the original tests did not fail.  I think it is a "lucky bug".  It's not, in general, guaranteed that a random forest will outperform a single decision tree.  It's very likely, but like I said, not guaranteed.  So I can safely set the tolerance at `0.9 * dtCorrect` and I have no problems when I run many trials.

zoq commented on src/mlpack/tests/random_forest_test.cpp:
```diff
@@ -483,4 +456,32 @@ BOOST_AUTO_TEST_CASE(RandomForestCategoricalTrainReturnEntropy)
+  bool success = false;
+  size_t trial = 0;
+
+  while (!success && trial < 10)
```
I think we should clarify the multiple trials here.

rcurtin replied:
Ah, 10 is too many I think, but the trials are still useful.  Comment fixed in cc7e2accb.

rcurtin commented:
I dug in to the problem a bit and exposed several other problems.  I fixed these, and when I finally got everything worked out:

 * Now there is a `--subspace_dim` option to specify the number of dimensions to use in a random forest.
 * Now there is a `--minimum_gain_split` option for random forests.
 * `--print_training_error` is deprecated and changed to `--print_training_accuracy` for the decision tree.
 * There is a very big efficiency improvement, although there is still more that could be done.  This was all I had time for now though.

@gmanlan @pensivevoice can you try this out and let me know if it helps?

rcurtin commented on src/mlpack/methods/random_forest/random_forest.hpp:
```diff
@@ -109,39 +128,54 @@ class RandomForest
+               const size_t minimumLeafSize = 1,
+               const double minimumGainSplit = 1e-7,
+               DimensionSelectionType dimensionSelector =
+                   DimensionSelectionType());
```
Actually that can be inferred by the compiler.  In your particular case you could just write

```
rf = RandomForest<>(dataset, labels,
    numClasses, numTrees, minimumLeafSize, 1e-7, MultipleRandomDimensionSelect(5));
```

rcurtin commented on src/mlpack/methods/decision_tree/multiple_random_dimension_select.hpp:
```diff
@@ -22,16 +22,30 @@ namespace tree {
  *
  * @tparam NumDimensions Number of random dimensions to select.
```
Thanks, fixed in d38d3115f.

zoq commented:
@mlpack-jenkins test this please

zoq commented on src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp:
```diff
@@ -50,47 +57,99 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+  {
+    if (UseWeights)
+    {
+      classWeightSums(sortedLabels[i], 0) += sortedWeights[i];
```
We could move this into the if/else case from above, to make this shorter.

rcurtin replied:
Yeah, agreed.  Done in 7b584bde.  I'm not sure it's actually shorter (didn't check too closely) but it's certainly easier to read.

zoq commented on src/mlpack/methods/decision_tree/decision_tree_impl.hpp:
```diff
@@ -552,9 +596,8 @@ double DecisionTree<FitnessFunction,
-  DimensionSelectionType dimensions(datasetInfo.Dimensionality());
-  for (size_t i = dimensions.Begin(); i != dimensions.End();
-       i = dimensions.Next())
+  for (size_t i = dimensionSelector.Begin(); i != dimensionSelector.End();
```
Maybe we should the same here as well: https://github.com/mlpack/ensmallen/pull/113#pullrequestreview-236418916

rcurtin replied:
Usually `End()` is just a `const size_t`, but I think paranoia is well warranted. :)  Done in 7b584bde.

rcurtin commented:
From my end this is ready to go, but I doubt the static code analysis checks are going to finish soon given the Jenkins build queue.  (I also don't believe anything will fail, so I don't think we should wait on them to finish.)  When it's merged I'll release 3.1.1 with this fix and any others that have been merged since the last release.

rcurtin commented on src/mlpack/methods/random_forest/random_forest_impl.hpp:
```diff
@@ -436,38 +456,46 @@ double RandomForest<
   #pragma omp parallel for reduction( + : avgGain)
   for (omp_size_t i = 0; i < numTrees; ++i)
   {
+    Timer::Start("bootstrap");
```
I just added these for my own debugging, but I decided to leave them in because they could be interesting.  If you run the binding with `--verbose`, it'll print all the timers collected during the run, which can be nice to figure out exactly where the time is going.  It may be more useful for developers than users, but I've always thought it's reasonably interesting to look at for users too. :)

gmanlan replied:
sounds good!

ShikharJ commented:
Thanks for the contribution, I'll review this today and try to get it merged at the earliest :+1:

rcurtin commented on src/mlpack/methods/random_forest/random_forest_main.cpp:
```diff
@@ -47,7 +47,12 @@ PROGRAM_INFO("Random forests",
     PRINT_PARAM_STRING("num_trees") +
-    " controls the number of trees in the random forest. If " +
+    " controls the number of trees in the random forest.  The " +
+    PRINT_PARAM_STRING("minimum_gain_split") + " parameter control the minimum "
```
Nice catch!  :+1: Are you okay if I merge this, or do you want to take a closer pass first?

ShikharJ replied:
Yeah, please go ahead. I'll probably take a look at how mlpack implements Random Forest Algorithm when I am faced with the situation of learning how Random Forests are supposed to work :smile: You might want to resolve the branch conflicts though.
